amr-wind/physics/ForestDrag.cpp

```diff
@@ -0,0 +1,118 @@
+#include "amr-wind/physics/ForestDrag.H"
+#include "amr-wind/CFDSim.H"
+#include "AMReX_iMultiFab.H"
+#include "AMReX_MultiFabUtil.H"
+#include "AMReX_ParmParse.H"
+#include "AMReX_ParReduce.H"
+#include "amr-wind/utilities/trig_ops.H"
+#include "amr-wind/utilities/IOManager.H"
+
+namespace amr_wind::forestdrag {
+
+namespace {} // namespace
+
+ForestDrag::ForestDrag(CFDSim& sim)
+    : m_sim(sim)
+    , m_repo(sim.repo())
+    , m_mesh(sim.mesh())
+    , m_velocity(sim.repo().get_field("velocity"))
+    , m_forest_drag(sim.repo().declare_field("forest_drag", 1, 1, 1))
+    , m_forest_blank(sim.repo().declare_field("forest_blank", 1, 1, 1))
+{
+    std::string forestfile("forest.amrwind");
+    amrex::ParmParse pp(identifier());
+    pp.query("forest_file", forestfile);
+    std::ifstream file(forestfile, std::ios::in);
+    if (!file.good()) {
+        amrex::Abort("Cannot find file " + forestfile);
+    }
+    //! TreeType xc yc height diameter cd lai laimax
+    amrex::Real value1, value2, value3, value4, value5, value6, value7, value8;
+    while (file >> value1 >> value2 >> value3 >> value4 >> value5 >> value6 >>
+           value7 >> value8) {
+        Forest f;
+        f.m_type_forest = value1;
+        f.m_x_forest = value2;
+        f.m_y_forest = value3;
+        f.m_height_forest = value4;
+        f.m_diameter_forest = value5;
+        f.m_cd_forest = value6;
+        f.m_lai_forest = value7;
+        f.m_laimax_forest = value8;
+        m_forests.push_back(f);
+    }
+    file.close();
+    m_sim.io_manager().register_output_var("forest_drag");
+    m_sim.io_manager().register_output_var("forest_blank");
+}
+
+void ForestDrag::post_init_actions()
+{
+    BL_PROFILE("amr-wind::" + this->identifier() + "::post_init_actions");
+    const auto& geom_vec = m_sim.repo().mesh().Geom();
+    const int nlevels = m_sim.repo().num_active_levels();
+    for (int level = 0; level < nlevels; ++level) {
+        const auto& geom = geom_vec[level];
+        const auto& dx = geom.CellSizeArray();
+        const auto& prob_lo = geom.ProbLoArray();
+        auto& velocity = m_velocity(level);
+        auto& drag = m_forest_drag(level);
+        auto& blank = m_forest_blank(level);
+        unsigned forestSize = m_forests.size();
+        for (unsigned ii = 0; ii < forestSize; ++ii) {
+            amrex::Real treelaimax = m_forests[ii].calc_lm();
+            const amrex::Real type_forest = m_forests[ii].m_type_forest;
```

Comment:
You shouldn't need all this. You can directly capture `Forest` struct to device. And then you can remove that DeviceVector of structs.

Reply:
I actually like the DeviceVector. So lets keep that.